src/util/side_metadata/global.rs

```diff
@@ -72,129 +77,252 @@ impl MappingState {
+/// * `global_metadata_spec_vec` - A vector of SideMetadataSpec objects containing all global side metadata.
 ///
-/// * `local_per_chunk` - The number of bytes of policy-specific side metadata required per chunk.
+/// * `local_metadata_spec_vec` - A vector of SideMetadataSpec objects containing all local side metadata.
 ///
 pub fn try_map_metadata_space(
     start: Address,
     size: usize,
-    global_per_chunk: usize,
-    local_per_chunk: usize,
+    global_metadata_spec_vec: &[SideMetadataSpec],
+    local_metadata_spec_vec: &[SideMetadataSpec],
 ) -> bool {
-    let mut aligned_start = start.align_down(BYTES_IN_CHUNK);
-    let aligned_end = (start + size).align_up(BYTES_IN_CHUNK);
-
-    // first chunk is special, as it might already be mapped, so it shouldn't be unmapped on failure
-    let mut munmap_first_chunk: Option<bool> = None;
+    debug_assert!(start.is_aligned_to(BYTES_IN_PAGE));
+    debug_assert!(size % BYTES_IN_PAGE == 0);
+
+    for spec in global_metadata_spec_vec {
+        // nearest page-aligned starting address
+        let mmap_start = address_to_meta_address(*spec, start).align_down(BYTES_IN_PAGE);
+        // nearest page-aligned ending address
+        let mmap_size = address_to_meta_address(*spec, start + size)
+            .align_up(BYTES_IN_PAGE)
+            .as_usize()
+            - mmap_start.as_usize();
+        if mmap_size > 0 {
+            // FIXME - This assumes that we never mmap a metadata page twice.
+            // While this never happens in our current use-cases where the minimum data mmap size is a chunk and the metadata ratio is larger than 1/64, it could happen if (min_data_mmap_size * metadata_ratio) is smaller than a page.
+            // E.g. the current implementation detects such a case as an overlap and returns false.
+            if !try_mmap_metadata(mmap_start, mmap_size).is_mappable() {
+                return false;
+            }
+        }
+    }
 
-    while aligned_start < aligned_end {
-        let res = try_mmap_metadata_chunk(aligned_start, global_per_chunk, local_per_chunk);
-        debug!(
-            "try_mmap_metadata_chunk({}, {:X}, {:X}) = {:?}",
-            aligned_start, global_per_chunk, local_per_chunk, res
-        );
-        if !res.is_mappable() {
-            if munmap_first_chunk.is_some() {
-                let mut munmap_start = if munmap_first_chunk.unwrap() {
-                    start.align_down(BYTES_IN_CHUNK)
-                } else {
-                    start.align_down(BYTES_IN_CHUNK) + BYTES_IN_CHUNK
-                };
-                // Failure: munmap what has been mmapped before
-                while munmap_start < aligned_start {
-                    ensure_munmap_metadata_chunk(munmap_start, global_per_chunk, local_per_chunk);
-                    munmap_start += SIDE_METADATA_PER_CHUNK;
-                }
+    let mut lsize: usize = 0;
+
+    for spec in local_metadata_spec_vec {
```

Comment:
Can you explain why we deal with 64/32bits differently? And why we only deal with them differently for local side metadata (not for the global as well)?

Reply:
For local side metadata, we always have to reserve address space for all local metadata required by all policies in MMTk, to be able to calculate a constant offset for each local metadata at compile-time (it's like assigning an ID to each policy).
As the plan is chosen at run-time, we will never know which subset of policies will be used during run-time.
We can't afford this much address space in 32-bits.
So, we switch to the chunk-based approach for this specific case.

The global metadata is different in that for each plan, we can calculate its constant base addresses at compile-time.
Using the chunk-based approach will need the same address space size as the current not-chunked approach.

Reply:
Thanks. Can you put those as comments in the code?